File: crm/src/main/java/serp/project/crm/ui/controller/OpportunityController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.crm.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import serp.project.crm.core.domain.dto.PageRequest;
import serp.project.crm.core.domain.dto.request.CreateOpportunityRequest;
import serp.project.crm.core.domain.dto.request.OpportunityFilterRequest;
import serp.project.crm.core.domain.dto.request.UpdateOpportunityRequest;
import serp.project.crm.core.domain.enums.OpportunityStage;
import serp.project.crm.core.usecase.ActivityUseCase;
import serp.project.crm.core.usecase.OpportunityUseCase;
import serp.project.crm.kernel.utils.AuthUtils;

@RestController
@RequestMapping("/api/v1/opportunities")
@RequiredArgsConstructor
@Slf4j
public class OpportunityController {

    private final OpportunityUseCase opportunityUseCase;
    private final ActivityUseCase activityUseCase;

    private final AuthUtils authUtils;

    @PostMapping
    public ResponseEntity<?> createOpportunity(@Valid @RequestBody CreateOpportunityRequest request) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        log.info("POST /api/v1/opportunities - Creating opportunity for tenant: {}", tenantId);
        var response = opportunityUseCase.createOpportunity(request, userId, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PatchMapping("/{id}")
    public ResponseEntity<?> updateOpportunity(
            @PathVariable Long id,
            @Valid @RequestBody UpdateOpportunityRequest request) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        log.info("PUT /api/v1/opportunities/{} - Updating opportunity for tenant: {}", id, tenantId);
        var response = opportunityUseCase.updateOpportunity(id, request, userId, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getOpportunityById(@PathVariable Long id) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        log.info("GET /api/v1/opportunities/{} - Fetching opportunity for tenant: {}", id, tenantId);
        var response = opportunityUseCase.getOpportunityById(id, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/{id}/activities")
    public ResponseEntity<?> getActivitiesByOpportunityId(
            @PathVariable Long id,
            @RequestParam(defaultValue = "1") Integer page,
            @RequestParam(defaultValue = "20") Integer size) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        log.info("GET /api/v1/opportunities/{}/activities - Fetching activities for tenant: {}, page: {}, size: {}",
                id, tenantId, page, size);

        PageRequest pageRequest = PageRequest.builder()
                .page(page)
                .size(size)
                .build();
        var response = activityUseCase.getActivitiesByOpportunity(id, tenantId, pageRequest);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping
    public ResponseEntity<?> getAllOpportunities(
            @RequestParam(defaultValue = "1") Integer page,
            @RequestParam(defaultValue = "20") Integer size) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        log.info("GET /api/v1/opportunities - Fetching all opportunities for tenant: {}, page: {}, size: {}",
                tenantId, page, size);

        PageRequest pageRequest = PageRequest.builder()
                .page(page)
                .size(size)
                .build();

        var response = opportunityUseCase.getAllOpportunities(tenantId, pageRequest);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/search")
    public ResponseEntity<?> filterOpportunities(@RequestBody(required = false) OpportunityFilterRequest request) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        OpportunityFilterRequest safeRequest = request != null ? request : OpportunityFilterRequest.builder().build();

        log.info("POST /api/v1/opportunities/search - Filtering opportunities for tenant: {}", tenantId);
        var response = opportunityUseCase.filterOpportunities(safeRequest, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PatchMapping("/{id}/stage")
    public ResponseEntity<?> changeStage(
            @PathVariable Long id,
            @RequestParam OpportunityStage newStage) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        log.info("PATCH /api/v1/opportunities/{}/stage - Changing stage to {} for tenant: {}", id, newStage,
                tenantId);
        var response = opportunityUseCase.changeOpportunityStage(id, newStage, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/{id}/close-won")
    public ResponseEntity<?> closeAsWon(@PathVariable Long id) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }
        log.info("POST /api/v1/opportunities/{}/close-won - Closing as won for tenant: {}", id, tenantId);
        var response = opportunityUseCase.closeOpportunityAsWon(id, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/{id}/close-lost")
    public ResponseEntity<?> closeAsLost(
            @PathVariable Long id,
            @RequestParam String lostReason) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        log.info("POST /api/v1/opportunities/{}/close-lost - Closing as lost for tenant: {}", id, tenantId);
        var response = opportunityUseCase.closeOpportunityAsLost(id, lostReason, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteOpportunity(@PathVariable Long id) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        log.info("DELETE /api/v1/opportunities/{} - Deleting opportunity for tenant: {}", id, tenantId);
        var response = opportunityUseCase.deleteOpportunity(id, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }
}

File: api_gateway/src/ui/controller/crm/opportunity_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	"github.com/serp/api-gateway/src/core/service/crm"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type OpportunityController struct {
	opportunityService crm.IOpportunityService
}

func (o *OpportunityController) CreateOpportunity(ctx *gin.Context) {
	var payload map[string]any
	if err := ctx.ShouldBindJSON(&payload); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.opportunityService.CreateOpportunity(ctx.Request.Context(), payload)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OpportunityController) UpdateOpportunity(ctx *gin.Context) {
	opportunityId, ok := utils.ValidateAndParseID(ctx, "opportunityId")
	if !ok {
		return
	}

	var payload map[string]any
	if err := ctx.ShouldBindJSON(&payload); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.opportunityService.UpdateOpportunity(ctx.Request.Context(), opportunityId, payload)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OpportunityController) GetOpportunityByID(ctx *gin.Context) {
	opportunityId, ok := utils.ValidateAndParseID(ctx, "opportunityId")
	if !ok {
		return
	}

	res, err := o.opportunityService.GetOpportunityByID(ctx.Request.Context(), opportunityId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OpportunityController) GetOpportunities(ctx *gin.Context) {
	page, size, valid := utils.ValidatePaginationParams(ctx)
	if !valid {
		return
	}

	res, err := o.opportunityService.GetOpportunities(ctx.Request.Context(), page, size)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OpportunityController) FilterOpportunities(ctx *gin.Context) {
	payload := map[string]any{}
	if ctx.Request.ContentLength != 0 {
		if err := ctx.ShouldBindJSON(&payload); err != nil {
			utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
			return
		}
	}

	res, err := o.opportunityService.FilterOpportunities(ctx.Request.Context(), payload)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OpportunityController) ChangeStage(ctx *gin.Context) {
	opportunityId, ok := utils.ValidateAndParseID(ctx, "opportunityId")
	if !ok {
		return
	}

	newStage := ctx.Query("newStage")
	if newStage == "" {
		utils.AbortErrorHandleCustomMessage(ctx, constant.GeneralBadRequest, "newStage query parameter is required")
		return
	}

	res, err := o.opportunityService.ChangeStage(ctx.Request.Context(), opportunityId, newStage)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OpportunityController) CloseAsWon(ctx *gin.Context) {
	opportunityId, ok := utils.ValidateAndParseID(ctx, "opportunityId")
	if !ok {
		return
	}

	res, err := o.opportunityService.CloseAsWon(ctx.Request.Context(), opportunityId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OpportunityController) CloseAsLost(ctx *gin.Context) {
	opportunityId, ok := utils.ValidateAndParseID(ctx, "opportunityId")
	if !ok {
		return
	}

	lostReason := ctx.Query("lostReason")
	if lostReason == "" {
		utils.AbortErrorHandleCustomMessage(ctx, constant.GeneralBadRequest, "lostReason query parameter is required")
		return
	}

	res, err := o.opportunityService.CloseAsLost(ctx.Request.Context(), opportunityId, lostReason)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OpportunityController) DeleteOpportunity(ctx *gin.Context) {
	opportunityId, ok := utils.ValidateAndParseID(ctx, "opportunityId")
	if !ok {
		return
	}

	res, err := o.opportunityService.DeleteOpportunity(ctx.Request.Context(), opportunityId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func NewOpportunityController(opportunityService crm.IOpportunityService) *OpportunityController {
	return &OpportunityController{opportunityService: opportunityService}
}
